include/exadg/functions_and_boundary_conditions/evaluate_functions.h

```diff
@@ -203,21 +224,137 @@ struct FunctionEvaluator<1, dim, Number>
+struct FunctionEvaluator<2, dim, Number>
+{
+  static inline DEAL_II_ALWAYS_INLINE //
+    dealii::Tensor<2, dim, dealii::VectorizedArray<Number>>
+    value(std::shared_ptr<dealii::Function<dim>>                      function,
+          dealii::Point<dim, dealii::VectorizedArray<Number>> const & q_points,
+          double const &                                              time)
+  {
+    dealii::Tensor<2, dim, dealii::VectorizedArray<Number>> value;
+
+    for(unsigned int d1 = 0; d1 < dim; ++d1)
+    {
+      for(unsigned int d2 = 0; d2 < dim; ++d2)
+      {
+        for(unsigned int v = 0; v < dealii::VectorizedArray<Number>::size(); ++v)
+        {
+          dealii::Point<dim> q_point;
+
+          for(unsigned int i = 0; i < dim; ++i)
+            q_point[i] = q_points[i][v];
+
+          function->set_time(time);
+
+          // Assuming row-major vector function equivalent
+          value[d1][d2][v] = function->value(q_point, d1 * dim + d2);
+        }
+      }
+    }
+
+    return value;
+  }
+
+  static inline DEAL_II_ALWAYS_INLINE //
+      dealii::Tensor<2, dim, dealii::VectorizedArray<Number>>
+      value(std::shared_ptr<FunctionCached<2, dim>> function,
+            unsigned int const                      face,
+            unsigned int const                      q,
+            unsigned int const                      quad_index)
+  {
+    dealii::Tensor<2, dim, dealii::VectorizedArray<Number>> value;
+
+    std::array<dealii::Tensor<2, dim, Number>, dealii::VectorizedArray<Number>::size()>
+      tensor_array;
+
+    for(unsigned int v = 0; v < dealii::VectorizedArray<Number>::size(); ++v)
+      tensor_array[v] = function->tensor_value(face, q, v, quad_index);
+
+    for(unsigned int d1 = 0; d1 < dim; ++d1)
+    {
+      for(unsigned int d2 = 0; d2 < dim; ++d2)
+      {
+        for(unsigned int v = 0; v < dealii::VectorizedArray<Number>::size(); ++v)
+          value[d1][d2][v] = tensor_array[v][d1][d2];
+      }
+    }
+
+    return value;
+  }
+
+  static inline DEAL_II_ALWAYS_INLINE //
+    dealii::SymmetricTensor<2, dim, dealii::VectorizedArray<Number>>
+    value_symmetric(std::shared_ptr<dealii::Function<dim>>                      function,
+                    dealii::Point<dim, dealii::VectorizedArray<Number>> const & q_points,
+                    double const &                                              time)
+  {
+    dealii::SymmetricTensor<2, dim, dealii::VectorizedArray<Number>> value;
+
+    for(unsigned int d1 = 0; d1 < dim; ++d1)
+    {
+      for(unsigned int d2 = d1; d2 < dim; ++d2)
+      {
+        for(unsigned int v = 0; v < dealii::VectorizedArray<Number>::size(); ++v)
+        {
+          dealii::Point<dim> q_point;
+
+          for(unsigned int i = 0; i < dim; ++i)
+            q_point[i] = q_points[i][v];
+
+          function->set_time(time);
+
+          // Assuming row-major vector function equivalent
+          value[d1][d2][v] = function->value(q_point, d1 * dim + d2);
```

Comment:
This line looks identical to the case `dealii::Tensor` above. Hence, I wonder whether we need this function. Again, my perspective is the user perspective and whether it is clear for a user which indices have to be implemented.

This does not have to be resolved here, I just want to raise this issue so that we keep this in mind for the future and potentially develop better solutions. In the end, the discussion seems to be about the design of `dealii::Function` and usability for application projects such as ExaDG. One should probably hide `VectorizedArray` content completely from pure users, but similar arguments would imply that the function interface visible in application examples uses `dealii::Tensor` and not scalar tensor components.

Reply:
I see what you mean. I also think that defining tensor components individually instead of returning the tensor directly is a worse experience for the user. Especially in cases like this where an assumption like row-major or column-major is made implicitly. So we should actually think about using `TensorFunction` wherever it makes sense without breaking the interface. I will definitely consider this point in future development.